x-pack/legacy/plugins/maps/public/reducers/map.js

```diff
@@ -359,55 +347,89 @@ function updateWithDataRequest(state, action) {
   return { ...state, layerList };
 }
 
+function findSourceDataRequestByLayerId(state, layerId) {
+  const layerDescriptor = findLayerById(state, layerId);
+  if (!layerDescriptor) {
+    return;
+  }
+  const index = layerDescriptor.__dataRequests.findIndex(dataRequest => {
+    return dataRequest.dataId === SOURCE_DATA_ID_ORIGIN;
+  });
+
+  return {
+    index: index,
+    dataRequest: layerDescriptor.__dataRequests[index]
+  };
+}
+
 
 function updateSourceDataRequest(state, action) {
-  const layerDescriptor = findLayerById(state, action.layerId);
-  if (!layerDescriptor) {
+  const { index, dataRequest } = findSourceDataRequestByLayerId(state, action.layerId);
+  if (index < 0) {
     return state;
   }
-  const dataRequest =   layerDescriptor.__dataRequests.find(dataRequest => {
-    return dataRequest.dataId === SOURCE_DATA_ID_ORIGIN;
+  return resetDataRequest(state, action, dataRequest, index, {
+    data: action.newData
   });
+}
+
+function clearDataRequest(state, action) {
+  const { dataRequest, index } = getValidDataRequest(state, action, true);
   if (!dataRequest) {
     return state;
   }
-
-  dataRequest.data = action.newData;
-  return resetDataRequest(state, action, dataRequest);
+  return resetDataRequest(state, action, dataRequest, index, {
+    data: null
+  });
 }
 
-
 function updateWithDataResponse(state, action) {
-  const dataRequest = getValidDataRequest(state, action);
-  if (!dataRequest) { return state; }
+  const { dataRequest, index } = getValidDataRequest(state, action, true);
+  if (!dataRequest) {
+    return state;
+  }
 
-  dataRequest.data = action.data;
-  dataRequest.dataMeta = { ...dataRequest.dataMetaAtStart, ...action.meta };
-  dataRequest.dataMetaAtStart = null;
-  return resetDataRequest(state, action, dataRequest);
+  return resetDataRequest(state, action, dataRequest, index, {
+    data: action.data,
+    dataMeta: { ...dataRequest.dataMetaAtStart, ...action.meta }
+  });
 }
 
-function resetDataRequest(state, action, request) {
-  const dataRequest = request || getValidDataRequest(state, action);
-  if (!dataRequest) { return state; }
+function resetDataRequest(state, action, dataRequest, index, newProps) {
+
+  const newDataRequest = {
+    ...dataRequest,
+    dataRequestToken: null,
+    dataId: action.dataId,
+    ...newProps
+  };
+
+  const layerDescriptor = findLayerById(state, action.layerId);
+  layerDescriptor.__dataRequests[index] = newDataRequest;
```

Comment:
This is still mutating state. Need to use spread on __dataRequests and then set index